Adding TargetInfo, Pixel Banner, AppEventListener

atrope opened the pull request:
Description of changes:
**This is a non-breaking change.**
Added Free Key-Value targeting to Android Banners(iOS SDK does not allow us to do that)
Added Pixel Banner (1x1)
Added the Ability to listen for Mobile App Events coming from AdMob/AdManager, for example in our case, we use to make custom Ads coming from AdManager with the Creative:

```
<script src='//media.admob.com/api/v1/google_mobile_app_ads.js'></script>
<script> 
     var clicktrack = '%%CLICK_URL_UNESC%%',image = 'https://XXXXXX,
    data = {image: image,clicktrack: clicktrack,};
document.addEventListener('DOMContentLoaded', function() {
    admob.events.dispatchAppEvent('data', JSON.stringify(data))
});</script>
```
This way we can pass any data through AdMob/AdManager and our ads.
https://developers.google.com/android/reference/com/google/android/gms/ads/doubleclick/AppEventListener?hl=pt-br


Example how to use Target and the new AppEventListener:
```

AdmobBanner(
        adUnitId: widget.adUnit,
        targetInfo: {'age': 29, 'domain': 'google'},
        adSize: AdmobBannerSize.PIXEL,
        listener: (AdmobAdEvent event, Map<String, dynamic> args) {
          switch (event) {
            case AdmobAdEvent.loaded:
              debugPrint('New Admob Ad loaded!');
              break;
            case AdmobAdEvent.appEvent:
              debugPrint("New App Event Name: ${args["name"]}");
              debugPrint("New App Event Data: ${args["data"]}");
              break;
            case AdmobAdEvent.failedToLoad:
              debugPrint('Admob failed to load. :( $args');
              break;
            default:
          }

```

```
I/flutter (23506): App Event Name: data
I/flutter (23506): App Event Data: {"image":"https://XXXXXX","clicktrack":"https://adclick.g.doubleclick.net/pcs/click?xaixxx...."}
```


That's it :), a Quick PR that should solve some issues.. I Intend also in the near future enable multiple ad sizes to the request (example we can request 320x50 and 320x100 in the same adunit..)


PS: I Never used AdMob, i don't know if we can send key/values like we send to AdManager/DFP, if anyone could test this would be great, if this does not work, we can have free target (for DFP/AdManager) and admob Target(Mobile Target Info like the official package) let me know what you think.

kmcgill88 commented on android/src/main/kotlin/com/shatsy/admobflutter/AdmobBanner.kt:
```diff
@@ -5,6 +5,9 @@ import android.view.View
 import com.google.android.gms.ads.AdView
+import com.google.android.gms.ads.doubleclick.AppEventListener
+import com.google.android.gms.ads.doubleclick.PublisherAdRequest
+import com.google.android.gms.ads.doubleclick.PublisherAdView
```
I've only worked with AdSense and Admob. How does DoubleClick fit into the family? I don't see a gradle dependency addition. I presume these classes are part of the same SDK?

atrope replied:
Hey @kmcgill88 Doubleclick is AdManager(Former DFP), the SDK is the same. DFP gives us a little bit more power and flexibility than AdMob and also We use same adserver for App/Web.

Speaking with our Google AM she said that we can use AdMob libraries for requesting AdManager AdUnits.

So I guess that we can Use Those views to also request AdMob.

kmcgill88 commented on android/src/main/kotlin/com/shatsy/admobflutter/AdmobBanner.kt:
```diff
@@ -14,18 +17,27 @@ import io.flutter.plugin.platform.PlatformView
 class AdmobBanner(context: Context, messenger: BinaryMessenger, id: Int, args: HashMap<*, *>?) : PlatformView, MethodCallHandler {
   private val channel: MethodChannel = MethodChannel(messenger, "admob_flutter/banner_$id")
-  private val adView: AdView = AdView(context)
+  private val adView: PublisherAdView = PublisherAdView(context)
```
How does this change affect existing AdMob users? Does iOS have the equivalent class?

atrope replied:
on iOS there is a DFPBannerView.

I Don't have adMob account to test, so if you or someone could try this implementation and check if we can still load ads from admob would be great.

Both AdView and PublisherAdView extends ViewGroup
https://developers.google.com/android/reference/com/google/android/gms/ads/doubleclick/PublisherAdView
https://developers.google.com/android/reference/com/google/android/gms/ads/AdView


with PublisherAdView we have a little more power with our requests. For example we  can set more than one size to request to our adserver while with the simple AdView we can't

kmcgill88 commented on lib/src/admob_banner_size.dart:
```diff
@@ -15,10 +15,16 @@ class AdmobBannerSize {
-  AdmobBannerSize.ADAPTIVE_BANNER({@required int width}):
-      width = width, height = -2, name = 'ADAPTIVE_BANNER';
+  static const AdmobBannerSize PIXEL =
+      AdmobBannerSize(width: 1, height: 1, name: 'PIXEL');
```
Sorry, I'm not following why you are setting a banner size of 1x1?

atrope replied:
As Stated in the PR, we use 1x1 AdUnits to pass messages and data to build custom ads experience within our application.
So one  example we do is that i can pass data to change for example the color of my background when the ad is loaded.
We don't need to render anything through the view so we use the smallest possible unit :)

kmcgill88 commented:
@atrope I've been looking into Ad Manager a little more to educate myself more about the product. Sounds like the two SDK's are extremely similar as you've mentioned. Considering this package has `Admob` in the name my first instinct is that if we want to add `AdManager` to do so in a vertical slice. By that I mean `AdmobBanner` remains unchanged and create `AdManagerBanner` and the related native implementations independently. I'm concerned mixing the two is just going to cause confusion for all the users who only use Admob.

@EliasDeuss @YoussefKababe what are your thoughts?

EliasDeuss commented:
@kmcgill88 We could mix it we would just have to separate them in the docs. It could be useful as long as its just documented well that its not really an Admob function.

kmcgill88 commented:
@EliasDeuss I still think the two should be separate.

EliasDeuss commented:
@kmcgill88 Did a little more research into is and I think you're right. I think that is should be its own doubleclick flutter package. Reduces the admob_flutter package size and confusion from users (and bugs).

atrope commented:
> @kmcgill88 Did a little more research into is and I think you're right. I think that is should be its own doubleclick flutter package. Reduces the admob_flutter package size and confusion from users (and bugs).

Package size would be the same.. this is 10 Lines of code added. 

Maybe we can refactor into one abstract class and make one Admob and one Admanager implementing it because the codes are basically the same but AdManager(and PublisherAdView) gives users more powerfull key values control and multiple ad sizes requests..

atrope commented:
Ok, so let's organize here :)

I've decoupled the common things and created abstract class `Banner` with this I created the new `AdmanagerBanner` which has all the things I wrote here before.
`AdmobBanner` continues with the same behaviour and almost the same code(Which now uses Banner class).

On iOS i did the same but the codes(for now) are the same. I Will later change the Admanager one to use DFPBannerView, but this PR was already big enough.

While i was modifying the classes i added the contentUrl option that gives users a way to add context to their apps pages
https://support.google.com/admanager/answer/6270563

Also I've upgraded to Android V2 Embedding (and removed completely V1) thus making this a breaking change. We can put that back but we should consider that with Flutter 1.22, flutter team deprecated V1.


Let me know what you think and i am sorry for this huge PR.

atrope commented:
@kmcgill88 @EliasDeuss

atrope commented:
One Last comment, in the end i ended up doing the iOS division also (DFPBannerView x GADBannerView). 
So now on iOS we also make different requests for each network. And all changes made to Android are also on iOS( Free Key-Value targeting that initially i said we couldn't do on iOS because of GADBannerView)

Let me know if you want this PR or not. 
If you think it is not. related to this plugin (I Really think it is) i will merge into my master and keep the aditions there instead of bringing it here :)

atrope commented:
I Will be closing this PR as I am migrating to another plugin for help and development,
The branch will be there if anyone needs those features:
```

  admob_flutter:
    git:
      url: https://github.com/SuaMusica/admob_flutter.git
      ref: feature/pixelBannerTarget
```
